drivers/pppou.c

```diff
@@ -0,0 +1,627 @@
+		pppou_do_rx(state);
+
+		log_debug("pppapi_close");
+		pppapi_close(state->ppp, 0);
+
+		mutexLock(state->lock);
+		log_debug("waiting for close to complete");
+		while (state->conn_state != CONN_STATE_DISCONNECTED) {
+			condWait(state->cond, state->lock, 0);
+			log_debug("still waiting for close to complete");
+		}
+		mutexUnlock(state->lock);
+
+		log_debug("connection closed");
+
+		serial_close(state->fd);
+		state->fd = -1;
+		state->conn_state = CONN_STATE_DISCONNECTED;
+
+		sleep(PPPOU_CONNECT_RETRY_SEC);
+	}
+
+	/* NOTE: never tested */
+	if (state->ppp) {
+		pppapi_close(state->ppp, 0);
+		pppapi_free(state->ppp);
+	}
+
+	endthread();
+}
+
+
+static int pppou_netifUp(pppou_priv_t *state)
+{
+	mutexLock(state->lock);
+	state->want_connected = 1;
+	condSignal(state->cond);
+	mutexUnlock(state->lock);
+
+	return 0;
+}
+
+
+static int pppou_netifDown(pppou_priv_t *state)
+{
+	/* Unconditional use of pppapi_close() in the status callback
+	 * can (and will) cause recursive firing of the callback */
+
+	mutexLock(state->lock);
+	state->conn_state = CONN_STATE_DISCONNECTING;
+	state->want_connected = 0;
+	mutexUnlock(state->lock);
+
+	return 0;
+}
+
+
+static pppou_priv_t *pppou_netifState(struct netif *netif)
+{
+	struct netif_alloc *s = (void *)netif;
+	pppou_priv_t *state = (void *) ((char *)s + ((sizeof(*s) + (_Alignof(pppou_priv_t) - 1)) & ~(_Alignof(pppou_priv_t) - 1)));
+	return state;
+}
+
+
+static void pppou_statusCallback(struct netif *netif)
+{
+	pppou_priv_t *state = pppou_netifState(netif);
+
+	if (netif->flags & NETIF_FLAG_UP) {
+		if (pppou_netifUp(state))
+			netif->flags &= ~NETIF_FLAG_UP;
+	}
+	else if (pppou_netifDown(state)) {
+		netif->flags |= NETIF_FLAG_UP;
+	}
+}
+
+
+static char *cfg_get_next_arg(char *arg)
+{
+	if (arg == NULL || *arg == '\0')
+		return NULL;
+
+	for (; *arg; arg++) {
+		if (*arg == ':') {
+			*arg++ = '\0';
+			break;
+		}
+	}
+
+	return arg;
+}
+
+
+static int pppou_netifInit(struct netif *netif, char *cfg)
+{
+	char *next;
+	speed_t speed;
+	pppou_priv_t* state;
+	int retries, flags = 0;
+
+	if (cfg == NULL || *cfg == '\0') {
+		log_error("no config");
+		return ERR_IF;
+	}
+
+	/*
+	 * NOTE: netif->state cannot be used to keep our private state as it is used
+	 * by LWiP PPP implementation, pass it as *ctx to callbacks
+	 */
+	state = netif->state;
+	netif->state = NULL;
+
+	memset(state, 0, sizeof(pppou_priv_t));
+	state->netif = netif;
+
+	for (; (next = cfg_get_next_arg(cfg)); cfg = next) {
+		if (!strncmp(cfg, "/dev/", 5)) {
+			state->serial_dev = cfg;
+			continue;
+		}
+
+		if ((speed = serial_speed_from_string(cfg)) >= 0) {
+			state->serial_speed = speed;
+			log_info("serial speed: %s bps => %d", cfg, state->serial_speed);
+			continue;
+		}
+
+		if (strcmp(cfg, "up") == 0) {
+			flags |= CFG_FLAG_DEFAULT_UP;
+			continue;
+		}
+
+		/* TODO: extend with other flags */
+	}
+
+	if (!state->serial_dev) {
+		log_error("no device");
+		return ERR_IF;
+	}
+
+	if (!state->serial_speed) {
```

Comment:
```suggestion
	if (state->serial_speed < 0) {
```
Function `serial_speed_from_string(cfg))` returns `-1` if  it dosn't find string.

Reply:
I don't agree. See line 543 where it's actually checked  `speed >= 0` and if true then `state->serial_speed = speed` is set, so until `serial_speed_from_string()` doesn't find proper speed `state->serial_speed` is set to zero (with memset), I assume zero as incorrect value because of `B0` value (`termios.h`).
The zero baud rate, `B0`, is used to terminate the connection.

Reply:
Ok. You're right.